server/src/main/java/org/elasticsearch/action/bulk/TransportShardBulkAction.java

```diff
@@ -114,157 +111,232 @@ protected boolean resolveIndex() {
+                listener.onResponse(
+                    new WritePrimaryResult<>(context.getBulkShardRequest(), context.buildShardResponse(), context.getLocationToSync(),
+                        null, context.getPrimary(), logger));
+            }
 
-    /** Executes bulk item requests and handles request execution exceptions */
-    static void executeBulkItemRequest(BulkPrimaryExecutionContext context, UpdateHelper updateHelper, LongSupplier nowInMillisSupplier,
-                                       MappingUpdatePerformer mappingUpdater, CheckedRunnable<Exception> waitForMappingUpdate)
-        throws Exception {
-        final DocWriteRequest.OpType opType = context.getCurrent().opType();
+        }.doRun();
+    }
 
-        final UpdateHelper.Result updateResult;
-        if (opType == DocWriteRequest.OpType.UPDATE) {
-            final UpdateRequest updateRequest = (UpdateRequest) context.getCurrent();
-            try {
-                updateResult = updateHelper.prepare(updateRequest, context.getPrimary(), nowInMillisSupplier);
-            } catch (Exception failure) {
-                // we may fail translating a update to index or delete operation
-                // we use index result to communicate failure while translating update request
-                final Engine.Result result = new Engine.IndexResult(failure, updateRequest.version(), SequenceNumbers.UNASSIGNED_SEQ_NO);
-                context.setRequestToExecute(updateRequest);
-                context.markOperationAsExecuted(result);
-                context.markAsCompleted(context.getExecutionResult());
-                return;
-            }
-            // execute translated update request
-            switch (updateResult.getResponseResult()) {
-                case CREATED:
-                case UPDATED:
-                    IndexRequest indexRequest = updateResult.action();
-                    IndexMetaData metaData = context.getPrimary().indexSettings().getIndexMetaData();
-                    MappingMetaData mappingMd = metaData.mappingOrDefault();
-                    indexRequest.process(metaData.getCreationVersion(), mappingMd, updateRequest.concreteIndex());
-                    context.setRequestToExecute(indexRequest);
-                    break;
-                case DELETED:
-                    context.setRequestToExecute(updateResult.action());
-                    break;
-                case NOOP:
-                    context.markOperationAsNoOp(updateResult.action());
+    /**
+     * Executes bulk item requests and handles request execution exceptions.
+     * @return {@code true} if request completed on this thread and the listener was invoked, {@code false} if the request triggered
+     *                      a mapping update that will finish and invoke the listener on a different thread
+     */
+    static boolean executeBulkItemRequest(BulkPrimaryExecutionContext context, UpdateHelper updateHelper, LongSupplier nowInMillisSupplier,
```

Comment:
instead of having this method return a boolean and internally spawn off the mapping update, I wonder if it would be better to leave that to the caller, which will have to check after running this method whether `context.isInitial()` or `context.isMappingUpdate()` and then do the right thing. This would hopefully reduce the complexity of this method a bit, and not require an ActionListener on it.

Reply:
Yea that sounds better I think, I'll try that now :)

Reply:
I gave up on this :( I did my best to at least shorten the method a little by drying up exception handling in c5edc6e74ef759070684a8a65413e107451e8f51 but it's pretty tricky to do the mapping update outside the method because its failure handling needs access to at least `updateResult` and it seems that any solution and the mapping update calll requires the engine result `result` as well, so those two would have to be set on the context somehow or so to be able to handle the mapping update outside the method. That seems like it would just make things worse in terms of readability?

Reply:
Maybe we should just reduce the amount of abstraction here (it seems we only pass ` MappingUpdatePerformer mappingUpdater, Consumer<ActionListener<Void>> waitForMappingUpdate` to get testability) and handle the tests that rely on this differently to simplify things instead?